compiler/hash-ast/src/parse.rs

```diff
@@ -206,23 +252,51 @@ where
         &self,
         filename: impl AsRef<Path>,
         directory: impl AsRef<Path>,
-    ) -> ParseResult<Modules<'c>> {
+    ) -> (Result<(), Vec<ParseError>>, Modules<'c>) {
         let filename = filename.as_ref();
         let directory = directory.as_ref();
         let entry = EntryPoint::Module { filename };
-        let (_, modules) = self.parse_main(entry, directory)?;
-        Ok(modules)
+        let (state, modules) = self.parse_main(entry, directory);
+
+        // check if the parser returned an error whilst parsing, if so extract the errors and pass them upwards
+        if let Err(errors) = state {
+            return (Err(errors), modules);
+        }
+
+        // If we need to visualise the file... then do so after the parser has finished
+        // the whole tree.
+        if self.visualise {
+            for module in modules.iter() {
+                println!(
+                    "file \"{}\":\n{}",
+                    module.filename().display(),
+                    module.ast()
+                );
+            }
+        }
+
+        (Ok(()), modules)
     }
 
     fn parse_interactive(
         &self,
         contents: &str,
         directory: impl AsRef<Path>,
-    ) -> ParseResult<(AstNode<'c, BodyBlock<'c>>, Modules<'c>)> {
+    ) -> (
+        Result<AstNode<'c, BodyBlock<'c>>, Vec<ParseError>>,
+        Modules<'c>,
+    ) {
         let directory = directory.as_ref();
         let entry = EntryPoint::Interactive { contents };
-        let (interactive, modules) = self.parse_main(entry, directory)?;
-        Ok((interactive.unwrap(), modules))
+        let (result, modules) = self.parse_main(entry, directory);
+
+        // Ensure that this method always returns a body block or an
+        // error since it can never be `None`.
+        match result {
+            Ok(Some(block)) => (Ok(block), modules),
+            Ok(None) => unreachable!(),
```

Comment:
How do we know that it will always return a body block?

Reply:
If it is an entry point to interactive job, it guarantees that a block will be returned from `parse_main`

Reply:
Fair enough; it might be worth it to add a message to unreachable, like `"Non-body block parsed node for interactive input"`.